cns/restserver/internalapi.go

```diff
@@ -143,6 +146,68 @@ func (service *HTTPRestService) SyncNodeStatus(dncEP, infraVnet, nodeID string,
 	return
 }
 
+// SyncHostNCVersion will check NC version from NMAgent and save it as host NC version in container status.
+// If NMAgent NC version got updated, CNS will refresh the pending programming IP status.
+func (service *HTTPRestService) SyncHostNCVersion(ctx context.Context, channelMode string, lastUpdatedHostNCTimeStamp time.Time, forceMarkIPAvailableTimeRange time.Duration) {
+	var hostVersionNeedUpdateNcList []string
+	service.RLock()
+	for _, containerstatus := range service.state.ContainerStatus {
+		// Will open a separate PR to convert all the NC version related variable to int. Change from string to int is a pain.
+		hostNcVersion, err := strconv.Atoi(containerstatus.HostNCVersion)
+		if err != nil {
+			log.Errorf("Received err when change containerstatus.HostNCVersion %s to int, err msg %v", containerstatus.HostNCVersion, err)
+			return
+		}
+		dncNcVersion, err := strconv.Atoi(containerstatus.DncNCVersion)
+		if err != nil {
+			log.Errorf("Received err when change containerstatus.DncNCVersion %s to int, err msg %v", containerstatus.DncNCVersion, err)
+			return
+		}
+		// host NC version is the NC version from NMAgent, if it's smaller than
+		if hostNcVersion < dncNcVersion {
+			hostVersionNeedUpdateNcList = append(hostVersionNeedUpdateNcList, containerstatus.ID)
+		}
+	}
+	service.RUnlock()
+	if len(hostVersionNeedUpdateNcList) > 0 {
+		c1 := make(chan map[string]int, 1)
+		go func() {
+			c1 <- service.nmagentClient.GetNcVersionListWithOutToken(hostVersionNeedUpdateNcList)
+		}()
+		select {
+		case newHostNCVersionList := <-c1:
+			service.Lock()
+			if newHostNCVersionList == nil {
+				logger.Errorf("Can't get vfp programmed NC version list from url without token")
+				if channelMode == cns.CRD && time.Since(lastUpdatedHostNCTimeStamp) > forceMarkIPAvailableTimeRange {
+					service.MarkAllPendingProgrammingIpsAsAvailableUntransacted()
```

Comment:
I dont understand this change. If NMAgent is not returning any list, how does marking an IP to available will help? If we assign this IP to a pod, data path will still be broken, right?